RecoTracker/DisplacedRegionalTracking/plugins/DisplacedRegionSeedingVertexProducer.cc

```diff
@@ -0,0 +1,254 @@
+#include <list>
+#include <vector>
+#include <limits>
+#include <string>
+#include <atomic>
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+#include "FWCore/ParameterSet/interface/FileInPath.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+#include "DataFormats/Candidate/interface/VertexCompositeCandidate.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/Common/interface/View.h"
+#include "DataFormats/Math/interface/Vector3D.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+
+#include "PhysicsTools/TensorFlow/interface/TensorFlow.h"
+
+#include "RecoTracker/DisplacedRegionalTracking/interface/DisplacedVertexCluster.h"
+
+using namespace std;
+typedef DisplacedVertexCluster::Distance Distance;
+typedef DisplacedVertexCluster::DistanceItr DistanceItr;
+
+class DisplacedRegionSeedingVertexProducer : public edm::global::EDProducer<> {
+public:
+  DisplacedRegionSeedingVertexProducer(const edm::ParameterSet &);
+  ~DisplacedRegionSeedingVertexProducer() override;
+  void produce(edm::StreamID, edm::Event &, const edm::EventSetup &) const override;
+  static void fillDescriptions(edm::ConfigurationDescriptions &);
+
+private:
+  // clustering parameters
+  double rParam_;
+
+  // selection parameters
+  double minRadius_;
+  double discriminatorCut_;
+  vector<string> input_names_;
+  vector<string> output_names_;
+
+  edm::EDGetTokenT<reco::BeamSpot> beamSpotToken_;
+  edm::EDGetTokenT<edm::View<reco::VertexCompositeCandidate> > trackClustersToken_;
+
+  tensorflow::Session *session_;
+
+  double getDiscriminatorValue(const DisplacedVertexCluster &, const reco::BeamSpot &) const;
+};
+
+DisplacedRegionSeedingVertexProducer::DisplacedRegionSeedingVertexProducer(const edm::ParameterSet &cfg)
+    : rParam_(cfg.getParameter<double>("rParam")),
+      minRadius_(cfg.getParameter<double>("minRadius")),
+      discriminatorCut_(cfg.getParameter<double>("discriminatorCut")),
+      input_names_(cfg.getParameter<vector<string> >("input_names")),
+      output_names_(cfg.getParameter<vector<string> >("output_names")),
+      beamSpotToken_(consumes<reco::BeamSpot>(cfg.getParameter<edm::InputTag>("beamSpot"))),
+      trackClustersToken_(
+          consumes<edm::View<reco::VertexCompositeCandidate> >(cfg.getParameter<edm::InputTag>("trackClusters"))) {
+  unsigned nThreads = cfg.getParameter<unsigned>("nThreads");
+  string singleThreadPool = cfg.getParameter<string>("singleThreadPool");
```

Comment:
These parameters should be untracked, because they do not affect the physics results (right?).

But is the `singleThreadPool` even necessary? That overload of `tensorflow::setThreading()` just prints a warning and calls the two-argument overload
https://github.com/cms-sw/cmssw/blob/3fa9bc2b0cc042ccc096cb794d35a25f227c3529/PhysicsTools/TensorFlow/src/TensorFlow.cc#L22-L26

Is the `nThreads` intended to be used in production, or only in private tests?

Reply:
> These parameters should be untracked, because they do not affect the physics results (right?).

Indeed they do not affect physics, so I have made `nThreads` untracked.

> 
> But is the `singleThreadPool` even necessary? That overload of `tensorflow::setThreading()` just prints a warning and calls the two-argument overload
> 
> https://github.com/cms-sw/cmssw/blob/3fa9bc2b0cc042ccc096cb794d35a25f227c3529/PhysicsTools/TensorFlow/src/TensorFlow.cc#L22-L26

I had not realized this had been deprecated, so I have removed `singleThreadPool` altogether.

> 
> Is the `nThreads` intended to be used in production, or only in private tests?

It's just for private tests right now. I'm not sure how a value greater than one would interact with the threading in CMSSW, so I don't change it from the default value of one in the config file ultimately meant for production.